File: big data(3 project)/thema_7/thema7_2.py
import pandas as pd
import numpy as np
from sklearn.cluster import AgglomerativeClustering
from scipy.cluster.hierarchy import dendrogram, linkage
import matplotlib.pyplot as plt
import csv
from sklearn.preprocessing import StandardScaler
# Specify the path to your CSV file
csv_file_path = 'C:\\Users\\kwstasbenek\\Desktop\\giannakopoulo_8\\Tsagkarakis_3h_ergasia\\3hErgasia\\europe.csv'

# Read the CSV file with specified encoding (e.g., UTF-8) and header row
europe_df = pd.read_csv(csv_file_path, encoding='utf-8', header=0)  # first row column names 

# Remove leading whitespaces from column names, since there are in almost everyone
europe_df.columns = europe_df.columns.str.strip()

# Display the DataFrame and its columns
print(europe_df)
print("Column Names:", europe_df.columns)

# Set 'Country' column as the index
#inplace=True specifies that the modification the modification operation should be performed directly on the object in memory, without creating a new object. 
europe_df.set_index('Country', inplace=True)

# Extract country names and features
country_names = europe_df.index
# iloc is a method in pandas that is used for integer-location based indexing.
# The : symbol in the indexing represents "all," so [:, :] is selecting all rows and all columns from the DataFrame 
numeric_features = europe_df.iloc[:, :] 
# Convert features to numeric, non numeric items will be NaN 
numeric_features = numeric_features.apply(pd.to_numeric, errors='coerce')
# Drop any rows with NaN values
numeric_features = numeric_features.dropna()
#basically all this is done to create a new df containing the same data points as the original europe_df except for country which is string 

#normalizing features
scaler = StandardScaler()
numeric_features_scaled = scaler.fit_transform(numeric_features)
# Perform hierarchical clustering using AgglomerativeClustering
n_clusters = 10

cluster = AgglomerativeClustering(n_clusters=n_clusters, metric='euclidean', linkage='ward')#'ward' minimizes variance within clusters 
cluster_labels = cluster.fit_predict(numeric_features)

# Linkage matrix for dendrogram, contains indices of merged clusters as well as distance/dissimillarity between the merged clusters 
linkage_matrix = linkage(numeric_features_scaled, method='ward')

# Plot the dendrogram
plt.figure(figsize=(15, 15))
dendrogram(linkage_matrix, labels=country_names, orientation='top', distance_sort='descending')
plt.title('Hierarchical Clustering Dendrogram')
plt.xlabel('Country')   
plt.ylabel('Distance')
plt.show()

File: big data(3 project)/thema_7/thema_7_2.R
# Remove all objects from the workspace
rm(list = ls())

# Close all open graphical devices
graphics.off()

# Read the CSV file
europe_data <- read.csv("C:\\Users\\kwstasbenek\\Desktop\\giannakopoulo_8\\Tsagkarakis_3h_ergasia\\3hErgasia\\europe.csv", header = TRUE)

# Remove leading whitespaces from column names
names(europe_data) <- trimws(names(europe_data))
#save country names 
country_names <- europe_data$Country
# Set 'Country' column as the row names
rownames(europe_data) <- europe_data$Country
europe_data$Country <- NULL  # Remove 'Country' column

# Convert data to numeric, non-numeric items will be converted to NA
europe_data <- apply(europe_data, 2, as.numeric)

# Remove rows with NA values
europe_data <- na.omit(europe_data)

# Normalize features
scaled_data <- scale(europe_data)

# Perform hierarchical clustering using hclust
#distance matrix using dist(), specifying euclidean 
distance_matrix <- dist(scaled_data, method = "euclidean")
#"ward.D2" refers to the variant of the Ward's method used, which is a criterion for measuring the dissimilarity between two clusters.
#The "D2" in "ward.D2" stands for squared Euclidean distance.
hierarchical_cluster <- hclust(distance_matrix, method = "ward.D2")

# Plot the dendrogram with 'Country' values as labels
plot(hierarchical_cluster, hang = -1, cex = 0.8, main = "Hierarchical Clustering Dendrogram",
     xlab = "Country", ylab = "Distance", labels = country_names)

